### PR TITLE
LOK-2361 : Add Http Monitor

### PR DESCRIPTION
**Pull request by aurang-zeb313**
## Description
Build a HTTP service monitor and register it in minion.
Proto file for http plugin.

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-2361

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: minion/plugins/pom.xml [hunk 1]
```diff
@@ -42,6 +42,7 @@
     <module>features</module>
     <module>nodescan-plugin</module>
     <module>shell-commands</module>
+      <module>http-plugin</module>
```
Comment (fooker) [inline minion/plugins/pom.xml:45]: Formating

File: shared-lib/protobuf/src/main/proto/http.proto [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  string inetAddress = 3;
+  repeated int32 port = 4;
+  repeated Header headers = 5;
+  string response = 6;
+  string responseText = 7;
+  string url = 8;
+  int32 retry = 9;
+  int64 timeout = 10;
+  string userAgent= 11;
+  bool verbose = 12;
```
Comment (fooker) [inline shared-lib/protobuf/src/main/proto/http.proto:48]: I think we can skip these - a old debugging artifact.

**Review by fooker — CHANGES_REQUESTED** [review]
The proto is okay, now. `verbose` can be removed

File: minion/plugins/http-plugin/pom.xml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                                <descriptor>mvn:org.apache.karaf.features/framework/${karaf.version}/xml/features</descriptor>
+                                <descriptor>mvn:org.apache.karaf.features/standard/${karaf.version}/xml/features</descriptor>
+                                <descriptor>mvn:org.opennms.lokahi.minion/grpc-features/${project.version}/xml/features</descriptor>
+                                <descriptor>mvn:org.opennms.lokahi.minion/ipc-features/${project.version}/xml/features</descriptor>
+                                <descriptor>mvn:org.opennms.lokahi.minion/observability-features/${project.version}/xml/features</descriptor>
+                                <descriptor>mvn:org.opennms.lokahi.minion/plugin-features/${project.version}/xml/features</descriptor>
+                                <descriptor>file:${project.build.directory}/feature/feature.xml</descriptor>
+                            </descriptors>
+                            <features>
+                                <feature>icmp*</feature>
```
Comment (smunir-onms) [inline minion/plugins/http-plugin/pom.xml:138]: I think it should be <feature>http*</feature>

**Comment by aurang-zeb313** [comment]
> The proto is okay, now. `verbose` can be removed
Done

**Review by kashifsardar181 — APPROVED** [review]

**Comment by aurang-zeb313** [comment]
Screenshot from 2024-04-08 21-32-09
Screenshot from 2024-04-08 21-31-10

**Review by cgorantla — COMMENTED** [review]
Please fix minion build failure

**Comment by aurang-zeb313** [comment]
> Please fix minion build failure

Build issue resolved.

File: shared-lib/lokahi-common-utils/src/main/java/org/opennms/horizon/shared/utils/Base64.java [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+ * License.
+ */
+package org.opennms.horizon.shared.utils;
+
+/**
+ * Performs base 64 encoding and decoding on byte arrays.
+ *
+ * @author <A HREF="mailto:weave@oculan.com">Brian Weaver </A>
+ */
+public final class Base64 {
```
[the remark this answers is not in the document]
Comment (aurang-zeb313) [reply shared-lib/lokahi-common-utils/src/main/java/org/opennms/horizon/shared/utils/Base64.java:29]: Base64 is used to encode userName and password in httpMonitor

File: minion/plugins/http-plugin/pom.xml [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
```
[the remark this answers is not in the document]
Comment (aurang-zeb313) [reply minion/plugins/http-plugin/pom.xml]: Yes, it is needed for test cases.

File: minion/plugins/http-plugin/src/main/feature/feature.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<features name="minion-plugin-http-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.6.0">
+
+
+    <feature name="http-shared-libs" hidden="true">
+        <feature>aries-blueprint</feature>
+        <feature>grpc-contract</feature>
+        <bundle dependency="true">mvn:org.opennms.lokahi.shared/lokahi-common-logging/${project.version}</bundle>
+        <bundle dependency="true">mvn:org.opennms.lokahi.shared/lokahi-common-utils/${project.version}</bundle>
+        <bundle dependency="true">mvn:org.opennms.lib/org.opennms.lib.tracker/0.7</bundle>
```
[the remark this answers is not in the document]
Comment (aurang-zeb313) [reply minion/plugins/http-plugin/src/main/feature/feature.xml]: I need only grpc-contract

File: minion/plugins/http-plugin/src/main/java/org/opennms/horizon/minion/http/HttpMonitor.java [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+        }
+
+        public void closeConnection() {
+            try {
+                if (m_httpSocket != null) {
+                    m_httpSocket.close();
+                    m_httpSocket = null;
+                }
+            } catch (final IOException e) {
+                e.fillInStackTrace();
```
[the remark this answers is not in the document]
Comment (aurang-zeb313) [reply minion/plugins/http-plugin/src/main/java/org/opennms/horizon/minion/http/HttpMonitor.java]: Removed.

File: minion/plugins/http-plugin/src/main/java/org/opennms/horizon/minion/http/HttpMonitor.java [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+                HttpMonitor.LOG.warn("Error closing socket connection", e);
```
[the remark this answers is not in the document]
Comment (aurang-zeb313) [reply minion/plugins/http-plugin/src/main/java/org/opennms/horizon/minion/http/HttpMonitor.java]: Done

File: minion/plugins/http-plugin/src/main/java/org/opennms/horizon/minion/http/HttpMonitor.java [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+                t.nextToken();
+            }
+
+            int serverResponse = -1;
+            if (t.hasMoreTokens()) {
+                try {
+                    serverResponse = Integer.parseInt(t.nextToken());
+                } catch (final NumberFormatException nfE) {
+
+                    HttpMonitor.LOG.info(
```
[the remark this answers is not in the document]
Comment (aurang-zeb313) [reply minion/plugins/http-plugin/src/main/java/org/opennms/horizon/minion/http/HttpMonitor.java]: Done

**Review by cgorantla — COMMENTED** [review]
I think we should be able to resolve just with url for any http service. We may not need ipaddress/hostname here. Since this was derived from OpenNMS Classic, we still have this code. we can improve upon this in next PR.

**Pull request merged**